### ISSUE TITLE
clang 15 regressions in translate-c test cases due to a warning promoted to error

### ISSUE THREAD

**Issue by andrewrk** [ziglang/zig#12682]
This issue regards the `llvm15` branch, with LLVM 15 rc3.

#### Function to integral

```c
#include <stdint.h>
int main(void) {
#if defined(__UINTPTR_MAX__) && __has_include(<unistd.h>)
    uintptr_t x = main;
    x = (uintptr_t)main;
#endif
    return 0;
}
```

With clang 15, outputs:

```
test.c:4:15: incompatible pointer to integer conversion initializing 'uintptr_t' (aka 'unsigned long') with an expression of type 'int (void)'
```

#### implicit casts

```c
#include <stdbool.h>

void fn_int(int x);
void fn_f32(float x);
void fn_f64(double x);
void fn_char(char x);
void fn_bool(bool x);
void fn_ptr(void *x);

void call() {
    fn_int(3.0f);
    fn_int(3.0);
    fn_int('ABCD');
    fn_f32(3);
    fn_f64(3);
    fn_char('3');
    fn_char('\x1');
    fn_char(0);
    fn_f32(3.0f);
    fn_f64(3.0);
    fn_bool(123);
    fn_bool(0);
    fn_bool(&fn_int);
    fn_int(&fn_int);
    fn_ptr(42);
}
```

With clang 15, outputs:

```
test.c:24:12: incompatible pointer to integer conversion passing 'void (*)(int)' to parameter of type 'int'
test.c:25:12: incompatible integer to pointer conversion passing 'int' to parameter of type 'void *'
```

----

Alright so what's the move here? Disable the warning in the test harness? Adjust the input C files? cc @ehaas @Vexu

**Comment by Vexu** [comment]
Were those made into errors? I get the same messages as warnings with clang 14. Either way it's probably safe to assume that cases like these exist plenty in C code and we should keep testing for them.

**Comment by andrewrk** [comment]
gotcha - I will look into this, thanks!

**Comment by andrewrk** [comment]
oh, to answer the question, yes they were made into an error in clang 15:

```
[nix-shell:~/dev/zig/build]$ ~/local/llvm14-release/bin/clang-14 -c test.c
test.c:24:12: warning: incompatible pointer to integer conversion passing 'void (*)(int)' to parameter of type 'int' [-Wint-conversion]
    fn_int(&fn_int);
           ^~~~~~~
test.c:3:17: note: passing argument to parameter 'x' here
void fn_int(int x);
                ^
test.c:25:12: warning: incompatible integer to pointer conversion passing 'int' to parameter of type 'void *' [-Wint-conversion]
    fn_ptr(42);
           ^~
test.c:8:19: note: passing argument to parameter 'x' here
void fn_ptr(void *x);
                  ^
2 warnings generated.

[nix-shell:~/dev/zig/build]$ ~/local/llvm15-release/bin/clang-15 -c test.c
test.c:24:12: error: incompatible pointer to integer conversion passing 'void (*)(int)' to parameter of type 'int' [-Wint-conversion]
    fn_int(&fn_int);
           ^~~~~~~
test.c:3:17: note: passing argument to parameter 'x' here
void fn_int(int x);
                ^
test.c:25:12: error: incompatible integer to pointer conversion passing 'int' to parameter of type 'void *' [-Wint-conversion]
    fn_ptr(42);
           ^~
test.c:8:19: note: passing argument to parameter 'x' here
void fn_ptr(void *x);
                  ^
2 errors generated.
```

**Comment by ehaas** [comment]
Since it's testing implicit casts which are no longer allowed, it seems reasonable to me to just remove the offending lines from the tests. These implicit conversions have technically been illegal since at least C99 so I guess they decided it's ok to make them errors now.

**Comment by andrewrk** [comment]
[the change is intentional and documented in the release notes](https://github.com/llvm/llvm-project/blob/release/15.x/clang/docs/ReleaseNotes.rst)

**Comment by andrewrk** [comment]
Fixed in ea4ef832e364d48b4670c71d948782df12e5e4ca.
